### ISSUE TITLE
Ability to update all Titles and URLs for a given Tripal Content Type.

### ISSUE THREAD

**Issue by laceysanderson** [tripal/tripal#47]
Currently if you change the page title format in the entity edit form, you have to edit each and every single entity/page of that type to see the new title take effect. Since Tripal Content often has a large number of pages, this isn't feasible. We should add a Tripal Job which can be scheduled from the entity edit page that will programmatically update each entity/page to the new title.

This is a feature that has been requested by multiple groups for Tripal 2 so we should try to get it into Tripal 3.

**Comment by shawnawsu** [comment]
I'm concerned about creating 404s if altering url structures and retroactively adjusting urls and titles (if urls are built using the title value). So I'd like to implement this using 2 Drupal modules, Pathauto (https://www.drupal.org/project/pathauto) and Redirect (https://www.drupal.org/project/redirect). 

Any objections?

**Comment by laceysanderson** [comment]
Currently, titles and URLs are altered using two different tokenized strings. You can edit them on the Tripal Content Type edit page, which I think is an intuitive place for Tripal Admin. 

If we switch to using Pathauto and redirect, will we be able to keep the same configuration section on the entity edit page?
Also, will we be able to keep the same tokens? (I assume this one is less an issue because we can just expose our own tokens using the API)

This does add an additional 2 dependencies to Tripal so I don't think it's a step that should be taken lightly... Furthermore, in the past we've greatly minimized the chance of 404's by restricting the URL alias' to fields that are required in chado. The bigger concern was non-unique URL alias' and while we made it very clear to admin that they had to choose something unique, there was still a risk.

All that said, I'm definitely not against the proposition. Can you explain your design a bit more? Do you see us using Pathauto to make the alias' and Redirect to redirect from the original URL to the new one? Will URLs in views be re-written automatically? Can this approach be used to ensure that old bookmarks to previous URL alias' redirect to the new alias (something our old approach didn't handle)?

**Comment by shawnawsu** [comment]
So I have one question before I try and answer your last paragraph. What do you mean by this, 'in the past we've greatly minimized the chance of 404's by restricting the URL alias' to fields that are required in chado'? 

So I've done a bit of research before answering, so apologies for the delay when I know we have a deadline coming up!  Let me try and answer your questions:

_Can you explain your design a bit more?_  I'd like to install both modules and transition path maintenance to the pathauto module. We could maintain the interface in the content type with a small function to update the central pathauto settings when that value is changed in the content type. This would result in two places you could maintain the alias. 

_Do you see us using Pathauto to make the alias' and Redirect to redirect from the original URL to the new one? Will URLs in views be re-written automatically?_  Pathauto allows you to set patterns using tokens to apply to content types, this is an example of a pathauto I have used on other websites: [node:menu-link:parents:join-path]/[node:title]. For example you could configure a gene like this: gene/[TripalEntity:id] ([TripalEntity:id] is defined as The unique ID of the tripal content.) Redirect would be put in place to make sure no 404s are created when url aliases are changed, so yes, when a url alias is changed and bulk updates applied the redirect module will automatically create redirects between the old and new alias. I don't know if I understand the question about urls in views? 

_Can this approach be used to ensure that old bookmarks to previous URL alias' redirect to the new alias (something our old approach didn't handle)?_ Yup! And with literally 4 clicks, you can set it and forget it.

I am happy to walk you through what these modules look like on another site so you can see what the administrative ui looks like and the functionality they provide. I will say adding and requiring configuration of these two modules in the setup process will add time to initial setup, but once we get the install profiles up and running that won't be as much of a problem.

**Comment by shawnawsu** [comment]
@laceysanderson  Thoughts?

**Comment by laceysanderson** [comment]
Sorry! Thanks for the *bump* 👍 

tl;dr: You've convinced me --go for it :-)

I think it should be ok with views due to the redirecting. I'll test it once you have it implemented. I think maintaining the old interface is important so I'm glad we can do that easily. I don't need a walk-through but thanks :-) I did use Pathauto a long time ago so I'm sure I'll pick it up again once it's implemented -plus we shouldn't need to use it directly?

**Comment by shawnawsu** [comment]
@laceysanderson I've got a completed version of this working but wanted to let you know I've removed the pathauto dependency. After consideration it just was too much for what needed to be achieved. Instead I stuck within the existing setup and just added the redirect module (to avoid the 404s when aliases are updated). 

@spficklin Everything is in the 7.x-3.x-dashboard branch. It needs to be tested thoroughly because I made changes to the TripalBundleUIController and TripalEntityController files and given their important role I'd appreciate additional testing on top of what I've done. The redirect module needs to be installed but I have added that to the tripal dependency list.

**Comment by spficklin** [comment]
I've tested and it works great!  I made one change so that when clicking the bulk buttons that it will also save the changes to the bundle as well.  I was afraid someone would change the URL or Title format and not realize to make it stick they would have to first save the bundle then go back in and click the bulk update button.  So now the button does both.

**Comment by spficklin** [comment]
Forgot to mention, it's now merged into the dev branch.
